File: games/mafia/frontend/types/ws-events.ts
import type { Room } from "./room";

// ---- Server -> Client ----
// Mirrors app/schemas/ws_events.py on the backend, one interface per `type`.

export interface RoomStateEvent {
  type: "room_state";
  room: Room;
}

export interface PlayerConnectionChangedEvent {
  type: "player_connection_changed";
  player_id: string;
  connected: boolean;
}

export interface ErrorEvent {
  type: "error";
  code: string;
  message: string;
}

export interface PongEvent {
  type: "pong";
}

export interface KickedEvent {
  type: "kicked";
}

export interface RoleOut {
  key: string;
  display_name: string;
  team: string;
  description: string;
  acts_at_night: boolean;
}

export interface RoleAssignedEvent {
  type: "role_assigned";
  role: RoleOut;
}

export interface InvestigationResultEvent {
  type: "investigation_result";
  target_player_id: string;
  team: string;
}

export interface NightResultEvent {
  type: "night_result";
  eliminated_player_id: string | null;
}

export interface EliminationResultEvent {
  type: "elimination_result";
  eliminated_player_id: string | null;
}

export interface RoleRevealOut {
  player_id: string;
  role_key: string;
  role_display_name: string;
  team: string;
}

export interface GameOverEvent {
  type: "game_over";
  winning_team: string;
  roles: RoleRevealOut[];
}

export interface VoteCastEvent {
  type: "vote_cast";
  player_id: string;
  target_player_id: string;
}

export interface MafiaPickOut {
  player_id: string;
  target_player_id: string | null;
  locked: boolean;
}

export interface MafiaNightPicksEvent {
  type: "mafia_night_picks";
  picks: MafiaPickOut[];
}

export interface NightTimerStartedEvent {
  type: "night_timer_started";
  duration_seconds: number;
}

export type ServerEvent =
  | RoomStateEvent
  | PlayerConnectionChangedEvent
  | ErrorEvent
  | PongEvent
  | KickedEvent
  | RoleAssignedEvent
  | InvestigationResultEvent
  | NightResultEvent
  | EliminationResultEvent
  | GameOverEvent
  | VoteCastEvent
  | MafiaNightPicksEvent
  | NightTimerStartedEvent;

// ---- Client -> Server ----

export interface PingCommand {
  type: "ping";
}

export interface SetReadyCommand {
  type: "set_ready";
  ready: boolean;
}

export interface UpdateProfileCommand {
  type: "update_profile";
  display_name?: string;
  avatar?: string;
}

export interface KickPlayerCommand {
  type: "kick_player";
  target_player_id: string;
}

export interface LeaveRoomCommand {
  type: "leave_room";
}

export interface StartGameCommand {
  type: "start_game";
  conflict_resolution: "kill_any" | "no_kill";
}

export interface AdvancePhaseCommand {
  type: "advance_phase";
}

export interface NightActionCommand {
  type: "night_action";
  target_player_id: string;
}

export interface CastVoteCommand {
  type: "cast_vote";
  target_player_id: string;
}

export interface LockNightActionCommand {
  type: "lock_night_action";
}

export type ClientCommand =
  | PingCommand
  | SetReadyCommand
  | UpdateProfileCommand
  | KickPlayerCommand
  | LeaveRoomCommand
  | StartGameCommand
  | AdvancePhaseCommand
  | NightActionCommand
  | CastVoteCommand
  | LockNightActionCommand;

File: games/would-you-rather/backend/app/schemas/ws_events.py
from typing import Literal

from pydantic import BaseModel

from app.schemas.room import RoomOut


class RoomStateEvent(BaseModel):
    type: Literal["room_state"] = "room_state"
    room: RoomOut


class PlayerConnectionChangedEvent(BaseModel):
    type: Literal["player_connection_changed"] = "player_connection_changed"
    player_id: str
    connected: bool


class ErrorEvent(BaseModel):
    type: Literal["error"] = "error"
    code: str
    message: str


class PongEvent(BaseModel):
    type: Literal["pong"] = "pong"


class KickedEvent(BaseModel):
    type: Literal["kicked"] = "kicked"


class QuestionShownEvent(BaseModel):
    type: Literal["question_shown"] = "question_shown"
    question_number: int
    total_questions: int
    option_a: str
    option_b: str


class PlayerVotedEvent(BaseModel):
    type: Literal["player_voted"] = "player_voted"
    player_id: str


class VotesRevealedEvent(BaseModel):
    type: Literal["votes_revealed"] = "votes_revealed"
    votes: dict[str, str]


class GameOverEvent(BaseModel):
    type: Literal["game_over"] = "game_over"
    recap: list[dict]
